pkg/utils/permission.go

```diff
@@ -352,3 +356,117 @@ func ApplyRoleBinding(ctx context.Context, client rbacclientv1.RoleBindingsGette
 	actual, err := client.RoleBindings(requiredCopy.Namespace).Update(ctx, existingCopy, metav1.UpdateOptions{})
 	return actual, true, err
 }
+
+// TemplatePermissionConfigFunc returns a func that can grant permission for addon agent
+// that is deployed by addon template.
+// the returned func will create a rolebinding to bind the clusterRole/role which is
+// specified by the user, so the user is required to make sure the existence of the
+// clusterRole/role
+func TemplatePermissionConfigFunc(
+	addonName string,
+	addonClient addonv1alpha1client.Interface,
+	hubKubeClient kubernetes.Interface,
+) agent.PermissionConfigFunc {
+
+	return func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) error {
+		template, err := GetDesiredAddOnTemplate(addonClient, addon)
+		if err != nil {
+			return err
+		}
+		if template == nil {
+			return nil
+		}
+
+		for _, registration := range template.Spec.Registration {
+			switch registration.Type {
+			case addonapiv1alpha1.RegistrationTypeKubeClient:
+
+				kcrc := registration.KubeClient
+				if kcrc == nil {
+					continue
+				}
+
+				for _, pc := range kcrc.HubPermissions {
+					switch pc.Type {
+					case addonapiv1alpha1.HubPermissionsBindingCurrentCluster:
+						klog.V(5).Infof("Set hub permission for addon %s/%s, UID: %s, APIVersion: %s, Kind: %s",
+							addon.Namespace, addon.Name, addon.UID, addon.APIVersion, addon.Kind)
+
+						owner := metav1.OwnerReference{
+							// TODO: use apiVersion and kind in addon object, but now they could be empty at some unknown reason
+							APIVersion: "addon.open-cluster-management.io/v1alpha1",
+							Kind:       "ManagedClusterAddOn",
+							Name:       addon.Name,
+							UID:        addon.UID,
+						}
+						err := createPermissionBinding(hubKubeClient,
```

Comment:
note the createPermissionBinding does not perform well. It will introduce too many get request.